02.Version cache - docker cache build framework

Kalimuthu-Velappan opened the pull request:
During docker build, host files can be passed to the docker build through
docker context files. But there is no straightforward way to transfer
the files from docker build to host.

This feature provides a tricky way to pass the cache contents from docker
build to host. It tar's the cached content and encodes them as base64 format
and passes it through a log file with a special tag as 'VCSTART and VCENT'.

Slave.mk in the host, it extracts the cache contents from the log and stores them
in the cache folder. Cache contents are encoded as base64 format for
easy passing.

<!--
     Please make sure you've read and understood our contributing guidelines:
     https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

     ** Make sure all your commits include a signature generated with `git commit -s` **

     If this is a bug fix, make sure your description includes "fixes #xxxx", or
     "closes #xxxx" or "resolves #xxxx"

     Please provide the following information:
-->

#### Why I did it

#### How I did it

#### How to verify it

#### Which release branch to backport (provide reason below if selected)

<!--
- Note we only backport fixes to a release branch, *not* features!
- Please also provide a reason for the backporting below.
- e.g.
- [x] 202006
-->

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205

#### Description for the changelog
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

#### Link to config_db schema for YANG module changes
<!--
Provide a link to config_db schema for the table for which YANG model
is defined
Link should point to correct section on https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-yang-models/doc/Configuration.md
-->

#### A picture of a cute animal (not mandatory but encouraged)

Kalimuthu-Velappan commented:
/azpw run Azure.sonic-buildimage

mssonicbld commented:
/AzurePipelines run Azure.sonic-buildimage

Kalimuthu-Velappan commented:
/azpw run Azure.sonic-buildimage

mssonicbld commented:
/AzurePipelines run Azure.sonic-buildimage

Kalimuthu-Velappan commented:
/azpw run Azure.sonic-buildimage

mssonicbld commented:
/AzurePipelines run Azure.sonic-buildimage

Kalimuthu-Velappan commented:
/azpw run Azure.sonic-buildimage

mssonicbld commented:
/AzurePipelines run Azure.sonic-buildimage

adyeung commented:
@xumia @liushilongbuaa submitter has taken the time to split the original code PR to smaller submissions for review, pls help take a look

liushilongbuaa commented:
Makefile.cache provides the same feature.

Kalimuthu-Velappan commented:
> Makefile.cache provides the same feature.

Yes, it is an extension of the DPKG caching framework.

Kalimuthu-Velappan commented:
/azpw run Azure.sonic-buildimage

mssonicbld commented:
/AzurePipelines run Azure.sonic-buildimage

liushilongbuaa commented:
1. I don't know what you want to copy.
2. From pipeline result, folder target/vcache is only 3KB and cache.tgz is empty.

Kalimuthu-Velappan commented:
> 

It is just a framework to copy files from the docker builder to the host. Currently, In this patch, nothing has been added to this file. This PR has only the framework part of the version cache framework.

The following patches in this series will add files to it. 

https://github.com/sonic-net/sonic-buildimage/pull/12003
https://github.com/sonic-net/sonic-buildimage/pull/12004
https://github.com/sonic-net/sonic-buildimage/pull/12005

and more coming on.

liushilongbuaa commented:
Did you try https://docs.docker.com/engine/reference/commandline/build/#custom-build-outputs?

Kalimuthu-Velappan commented:
> Did you try https://docs.docker.com/engine/reference/commandline/build/#custom-build-outputs?

 tried the below commands, but it doesn't generate the expected out file in the host.

#cat Dockerfile
FROM debian:buster AS build-stage
RUN mkdir -p /test
RUN echo "MY CNT" > /test/myfile.txt

FROM scratch AS export-stage
COPY --from=build-stage /test /

#docker build --no-cache -o out .
#ls
Dockerfile

xumia commented on src/sonic-build-hooks/scripts/post_run_buildinfo:
```diff
@@ -1,12 +1,22 @@
+tar -C ${PKG_CACHE_PATH} --exclude=cache.tgz -zcvf ${PKG_CACHE_FILE_NAME} .
+#set +x
+if [[ ! ${IMAGENAME} =~ host-image ]]; then
+	sleep 1;echo  -e "\n_VCSTART_"; (cat ${PKG_CACHE_FILE_NAME} | base64); echo -e "_VCEND_\n";sleep 1
```
Can we use the scp command or the other RPC APIs to copy the content to the slave container? It is a little workaround to copy content by log.

Kalimuthu-Velappan commented:
Using SCP/RCP to copy the files from docker builder to host might not work in all the combinations. This creates a security and authentication problem on the multiuser environment even if we use public authentication keys. The sonic-slave docker image is built directly on the build server which requires scp access on the build server to copy the files.

The -o option is not working with image tag option. 
If we run with -o option, it only generates the output and it does not generate the docker image.

# DOCKER_BUILDKIT=1 docker build --no-cache  -t test:latest -o out .
 => exporting to client                                                                                                                                                                         
 => copying files 56B                                                                                                                                                                    

# find out
out
out/a.log
# docker images |grep "test *latest"
< --- NO test:latest IMAGE IS GENERATED --- >


When we run without -o option, the docker image is generated, but the output is empty. 

# DOCKER_BUILDKIT=1 docker build --no-cache  -t test:latest  .
 => exporting to image                                                                                                                                                                      
 => => exporting layers                                                                                                                                                                    
 => => writing image sha256:f1201b5415c4d7ad943f55f32e1a8102d9467da2c0f791ca822ce38c791f6a39                                                                                               
 => => naming to docker.io/library/test:latest                                                                                                                                            

# find out
< --- NO OUTPUT GENERATED --- >
# docker images |grep "test *latest"
test                                               latest              f1201b5415c4        12 seconds ago      28B

liushilongbuaa commented on Makefile.work:
```diff
@@ -225,7 +225,11 @@ SLAVE_TAG = $(shell \
 	$(SLAVE_BASE_IMAGE):$(SLAVE_BASE_TAG) \
-		target 
+		target \
+				$(SLAVE_BASE_IMAGE):$(SLAVE_BASE_TAG) \
```
This line is the same as 227.
When they will be different?

Kalimuthu-Velappan replied:
For the sonic-slave image, both are the same, but for the sonic dockers, the first one takes the docker name and the second one takes the docker image tag.
Eg:
+ DOCKER_IMAGE=docker-database
+ TARGET_PATH=target
+ DOCKER_IMAGE_TAG=docker-database-username:49592b425b7

liushilongbuaa replied:
Suggest adding parameters when it is used.
It is confusing in this PR.

Kalimuthu-Velappan replied:
Got it, will change this.

Kalimuthu-Velappan replied:
done

Kalimuthu-Velappan commented:
The below sequence solves both the docker image generation and intermediate cache file output to the host. But the only problem is that we need to run the docker build command three times.  Though we are reusing the image layers, still little overhead on the build time, and requires a lot of changes in the docker build process in the sonic.  Let me know your thought on this?

# cat Dockerfile
	# Base docker build
	FROM debian:buster as build
	RUN dd if=/dev/zero of=/root/cache.data bs=100M count=1

	# Copy the cache data to host
	From scratch as output
	COPY --from=sonic-image-tmp:latest /root/cache.data cache.data

	# Clean up the cache data
	FROM sonic-image-tmp:latest as final
	RUN rm /root/cache.data

# cat build.sh
	rm -rf out; docker rmi sonic-image:latest

	DOCKER_BUILDKIT=1 docker build  --no-cache --tag sonic-image-tmp:latest --target build .;
	DOCKER_BUILDKIT=1 docker build  --cache-from sonic-image-tmp:latest --target output -o out .;
	DOCKER_BUILDKIT=1 docker build  --no-cache --cache-from sonic-image-tmp:latest --target final --tag sonic-image:latest .
	docker rmi sonic-image-tmp:latest

	find out; docker images |grep "sonic-image *latest"


======================================================
# bash -x ./build.sh
+ rm -rf out
+ docker rmi sonic-image:latest
Untagged: sonic-image:latest
Deleted: sha256:7799604f2b5b5649aed23df18fdb20f9125f1d7d45279c583b1dee993379a322
+ DOCKER_BUILDKIT=1
+ docker build --no-cache --tag sonic-image-tmp:latest --target build .
 => CACHED [build 1/2] FROM docker.io/library/debian:buster                                                                                                                                                                     => [build 2/2] RUN dd if=/dev/zero of=/root/cache.data bs=100M count=1                                                                                                                                                         => => exporting layers                                                                                                                                                                                                         => => writing image sha256:db76b329801e567580cc45e294709855c443f6acab629cc89d804eea8867f99f                                                                                                                                    => => naming to docker.io/library/sonic-image-tmp:latest                                                                                                                                                                      


+ DOCKER_BUILDKIT=1
+ docker build --cache-from sonic-image-tmp:latest --target output -o out .
 => importing cache manifest from sonic-image-tmp:latest                                                                                                                                                                        => FROM docker.io/library/sonic-image-tmp:latest                                                                                                                                                                               => => resolve docker.io/library/sonic-image-tmp:latest                                                                                                                                                                         => CACHED [output 1/1] COPY --from=sonic-image-tmp:latest /root/cache.data cache.data                                                                                                                                          => exporting to client                                                                                                                                                                                                        
 => => copying files 104.88MB                                                                                                                                                                                                  

+ DOCKER_BUILDKIT=1 
+ docker build --no-cache --cache-from sonic-image-tmp:latest --target final --tag sonic-image:latest .
 => [internal] load metadata for docker.io/library/sonic-image-tmp:latest                                                                                                                                                       => importing cache manifest from sonic-image-tmp:latest                                                                                                                                                                        => CACHED [final 1/2] FROM docker.io/library/sonic-image-tmp:latest                                                                                                                                                            => [final 2/2] RUN rm /root/cache.data                                                                                                                                                                                         => exporting to image                                                                                                                                                                                                          => => exporting layers                                                                                                                                                                                                         => => writing image sha256:ad3df1f88f3af5353bebd157744a1b3eab294fead83e58497b6c57787011f8cc                                                                                                                                    => => naming to docker.io/library/sonic-image:latest                                                                                                                                                                          

+ docker rmi sonic-image-tmp:latest
Untagged: sonic-image-tmp:latest
Deleted: sha256:db76b329801e567580cc45e294709855c443f6acab629cc89d804eea8867f99f

===============OUTPUT=============
+ find out
out
out/cache.data

+ docker images | grep 'sonic-image *latest'
sonic-image                                        latest              ad3df1f88f3a        1 second ago        219MB

zhangyanzhao commented:
@xumia @liushilongbuaa can you please help to review the PR? Thanks.

Kalimuthu-Velappan commented:
/azpw run Azure.sonic-buildimage

mssonicbld commented:
/AzurePipelines run Azure.sonic-buildimage

xumia commented:
@Kalimuthu-Velappan ,  the solution is good.

Can we only use two docker build steps? Split the Dockerfile into two parts, Dockerfile1 and Dockerfile2, Dockerfile1 is to do the normal installation without cleanup, Dockerfile2 is only to do the cleanup, it can be generated from a common j2 template for all docker builds. We can add a step to copy the content from the image. It is not necessary to depend on DOCKER_BUILDKIT, and makes the copy steps more controllable not in the Dockerfile.

But I think the solution you provided is good enough.

Kalimuthu-Velappan commented:
> @Kalimuthu-Velappan , the solution is good.
> 
> Can we only use two docker build steps? Split the Dockerfile into two parts, Dockerfile1 and Dockerfile2, Dockerfile1 is to do the normal installation without cleanup, Dockerfile2 is only to do the cleanup, it can be generated from a common j2 template for all docker builds. We can add a step to copy the content from the image. It is not necessary to depend on DOCKER_BUILDKIT, and makes the copy steps more controllable not in the Dockerfile.
> 
> But I think the solution you provided is good enough.

cat Dockerfile

# Base docker build
FROM debian:buster
RUN dd if=/dev/zero of=/root/cache.data bs=100M count=1
RUN echo $(date)

cat Dockerfile.cleanup

# Base docker build
FROM sonic-image:latest

# Copy the cache data to host
From scratch as output
COPY --from=sonic-image:latest /root/cache.data cache.data

# Clean up the cache data
FROM sonic-image:latest as final
RUN rm /root/cache.data

cat build.sh

#!/bin/bash
set -x
rm -rf out; docker rmi sonic-image-tmp:latest sonic-image:latest

docker build  --no-cache --tag sonic-image:latest .;
docker tag sonic-image:latest sonic-image-tmp:latest
DOCKER_BUILDKIT=1 docker build -f Dockerfile.cleanup  --target output -o out .;
DOCKER_BUILDKIT=1 docker build -f Dockerfile.cleanup  --no-cache --target final --tag sonic-image:latest .
docker rmi sonic-image-tmp:latest

find out; docker images |grep "sonic-image.*latest"

OUTPUT:

    rm -rf out
    docker rmi sonic-image-tmp:latest sonic-image:latest
    Untagged: sonic-image:latest
    Deleted: sha256:7fbc4abec8831240c9d1c574285fdd11b5c47a758d9e957a647ec5271bc750b0
    Error: No such image: sonic-image-tmp:latest
    docker build --no-cache --tag sonic-image:latest .
    Sending build context to Docker daemon 8.192kB
    Step 1/3 : FROM debian:buster
    ---> 4a7a1f401734
    Step 2/3 : RUN dd if=/dev/zero of=/root/cache.data bs=100M count=1
    ---> Running in 0fe5524ba638
    1+0 records in
    1+0 records out
    104857600 bytes (105 MB, 100 MiB) copied, 0.129012 s, 813 MB/s
    Removing intermediate container 0fe5524ba638
    ---> 76fea604fe86
    Step 3/3 : RUN echo $(date)
    ---> Running in 4aa14234d98e
    Wed Nov 23 13:46:58 UTC 2022
    Removing intermediate container 4aa14234d98e
    ---> 07028c1daa48
    Successfully built 07028c1daa48
    Successfully tagged sonic-image:latest
    docker tag sonic-image:latest sonic-image-tmp:latest
    DOCKER_BUILDKIT=1
    docker build -f Dockerfile.cleanup --target output -o out .
    [+] Building 0.6s (5/5) FINISHED
    => [internal] load build definition from Dockerfile.cleanup 0.0s
    => => transferring dockerfile: 377B 0.0s
    => [internal] load .dockerignore 0.0s
    => => transferring context: 2B 0.0s
    => FROM docker.io/library/sonic-image:latest 0.0s
    => => resolve docker.io/library/sonic-image:latest 0.0s
    => CACHED [output 1/1] COPY --from=sonic-image:latest /root/cache.data cache.data 0.0s
    => exporting to client 0.3s
    => => copying files 104.88MB 0.3s
    DOCKER_BUILDKIT=1
    docker build -f Dockerfile.cleanup --no-cache --target final --tag sonic-image:latest .
    [+] Building 0.6s (6/6) FINISHED
    => [internal] load build definition from Dockerfile.cleanup 0.0s
    => => transferring dockerfile: 46B 0.0s
    => [internal] load .dockerignore 0.0s
    => => transferring context: 2B 0.0s
    => [internal] load metadata for docker.io/library/sonic-image:latest 0.0s
    => CACHED [final 1/2] FROM docker.io/library/sonic-image:latest 0.0s
    => [final 2/2] RUN rm /root/cache.data 0.6s
    => exporting to image 0.0s
    => => exporting layers 0.0s
    => => writing image sha256:b7337d033d1088956a563ea2b111f467ba015566b3afb9b07b0b5964113699d5 0.0s
    => => naming to docker.io/library/sonic-image:latest 0.0s
    docker rmi sonic-image-tmp:latest
    Untagged: sonic-image-tmp:latest
    Deleted: sha256:07028c1daa486d4adbcd1fb0ae4ebf634f80f0fba1c204fd766d918e9ccc8a52
    Deleted: sha256:76fea604fe863f93a94fac7b608ac794598769369601995bac5d0d777d6c9e5e
    find out
    out
    out/cache.data
    docker images
    grep 'sonic-image.*latest'
    sonic-image latest b7337d033d10 1 second ago 219MB

Kalimuthu-Velappan commented:
/azpw run Azure.sonic-buildimage

mssonicbld commented:
/AzurePipelines run Azure.sonic-buildimage

xumia commented on src/sonic-build-hooks/scripts/pre_run_buildinfo:
```diff
@@ -1,7 +1,11 @@
+
 . /usr/local/share/buildinfo/scripts/buildinfo_base.sh
 
+#set -x 
+
```
```suggestion
```

xumia commented on src/sonic-build-hooks/scripts/pre_run_buildinfo:
```diff
@@ -1,7 +1,11 @@
+
 . /usr/local/share/buildinfo/scripts/buildinfo_base.sh
 
+#set -x 
```
Could you please remove the unuse lines? Thanks.
Line 7, 18, 20.

```
#set -x
#set +x
```

Kalimuthu-Velappan replied:
done

xumia commented on src/sonic-build-hooks/scripts/pre_run_buildinfo:
```diff
@@ -11,14 +15,20 @@ mkdir -p $LOG_PATH
     cp -f $VERSION_DEB_PREFERENCE /etc/apt/preferences.d/
 fi
 
+DISTRO=${DISTRO} apt-get update && apt-get install -y rsync
+
```
We do not need it, right?

xumia commented on src/sonic-build-hooks/scripts/post_run_buildinfo:
```diff
@@ -1,12 +1,15 @@
 
-[ -d $BUILD_VERSION_PATH ] && [ ! -z "$(ls -A $BUILD_VERSION_PATH)" ] && cp -rf $BUILD_VERSION_PATH/* $POST_VERSION_PATH
-rm -rf $BUILD_VERSION_PATH/*
+	
```
Additional space characters can be removed.

Kalimuthu-Velappan replied:
done

xumia commented:
@Kalimuthu-Velappan , looks good to me, only left one comment, do we need to install rsync?

Kalimuthu-Velappan commented:
> 

rsync is required for go module sync and it is being used by other PRs. I will find better way to remove this and will on it. 
We can keep this for now, I will address this as part of other PR.

xumia commented:
@Kalimuthu-Velappan , could you please fix the code conflicts?

kv-y commented on src/sonic-build-hooks/scripts/post_run_buildinfo:
```diff
@@ -1,12 +1,14 @@
-[ -d $BUILD_VERSION_PATH ] && [ ! -z "$(ls -A $BUILD_VERSION_PATH)" ] && cp -rf $BUILD_VERSION_PATH/* $POST_VERSION_PATH
-rm -rf $BUILD_VERSION_PATH/*
+#Save the cache file for exporting it to host.
+tar -C ${PKG_CACHE_PATH} --exclude=cache.tgz -zcvf /cache.tgz .
```
@Kalimuthu-Velappan 

BTW, what do you think about using pigz here?
In my tests I get a big improvements in time when use it instead of gzip:
https://github.com/sonic-net/sonic-buildimage/pull/12825

Kalimuthu-Velappan replied:
Thanks for your suggestion. 
As you said, I have already used this in the binary optimization PR.

mssonicbld commented:
Cherry-pick PR to 202211: https://github.com/sonic-net/sonic-buildimage/pull/13771
